File: src/main/java/discretemath/combination/GrayCode.java
package discretemath.combination;

import discretemath.common.BitString;

public class GrayCode {

	private final BitString[] table;

	private GrayCode(BitString[] table) {
		this.table = table;
	}

	public static GrayCode forN(int n) {
		BitString[] bitSetTable = createBitStringTable(n);

		return new GrayCode(bitSetTable);
	}

	public BitString get(int index) {
		return table[index];
	}

	public int find(BitString bitString) {
		if (bitString.length() != getVariablesCount()) {
			throw new IllegalArgumentException(bitString.toString());
		}

		for (int i = 0; i < table.length; i++) {
			BitString string = table[i];
			if (bitString.equals(string)) {
				return i;
			}
		}

		throw new IllegalStateException();
	}

	public int getCombinationsCount() {
		return table.length;
	}

	public int getVariablesCount() {
		return table[0].length();
	}

	@Override
	public String toString() {
		StringBuilder builder = new StringBuilder();
		for (BitString bitString : table) {
			for (int i = 0; i < bitString.length(); i++) {
				boolean b = bitString.get(i);
				builder.append(b ? 1 : 0).append("\t");
			}
			builder.append("\n");
		}

		return builder.toString();
	}

	private static BitString[] createBitStringTable(int n) {
		int rowsCount = 1 << n;

		BitString[] table = new BitString[rowsCount];
		for (int i = 0; i < table.length; i++) {
			table[i] = BitString.forN(n);
		}

		for (int i = 0; i < rowsCount; i++) {
			int grayCode = i ^ (i >> 1);
			for (int j = 0; j < n; j++) {
				boolean value = ((grayCode >> j) & 1) == 1;
				table[i].set(n - 1 - j, value);
			}
		}

		return table;
	}
}

File: src/main/scala/discretemath/combination/GrayCode.scala
package discretemath.combination

import discretemath.common.BitString

class GrayCode private(val table: Array[BitString]) {
  def get(index: Int): BitString = table(index)

  def find(bitString: BitString): Int = {
    if (bitString.length != getVariablesCount) throw new IllegalArgumentException(bitString.toString)
    for (i <- table.indices) {
      val string = table(i)
      if (bitString == string) return i
    }
    throw new IllegalStateException
  }

  def getCombinationsCount: Int = table.length

  def getVariablesCount: Int = table(0).length

  override def toString: String = {
    val builder = new StringBuilder
    for (bitString <- table) {
      for (i <- 0 until bitString.length) {
        val b = bitString.get(i)
        builder.append(if (b) 1
        else 0).append("\t")
      }
      builder.append("\n")
    }
    builder.toString
  }
}

object GrayCode {
  def forN(n: Int): GrayCode = {
    val bitSetTable = createBitStringTable(n)
    new GrayCode(bitSetTable)
  }

  private def createBitStringTable(n: Int) = {
    val rowsCount = 1 << n
    val table = new Array[BitString](rowsCount)
    for (i <- table.indices) {
      table(i) = BitString.forN(n)
    }
    for (i <- 0 until rowsCount) {
      val grayCode = i ^ (i >> 1)
      for (j <- 0 until n) {
        val value = ((grayCode >> j) & 1) == 1
        table(i).set(n - 1 - j, value)
      }
    }
    table
  }
}
